### DISCUSSION TITLE
Releases and Breaking Changes: MSRV, `half 2.3` intrinsics, `exr 2.0.0`

### DISCUSSION THREAD

**Discussion by johannesvollmer** [johannesvollmer/exrs#217] [General]
Versioning is hard, let's discuss it. Some events that sparked discussion:
- `half` released a new version with intrinsic conversion operations, which should be a lot faster
- the new `half` version uses the newest Rust version, `1.70.0` in a minor release, which sparked a lot of debate on its own (https://github.com/starkat99/half-rs/issues/97)
- we prevented old projects failing by restricting the version of `half` to our own MSRV (#213), which excludes the new intrinsics, in a minor `exr` release

See #191, where a lot of previous discussion already happened.

Maybe some kind of strategy should established for this repository? Maybe I just need to get all the details right about how to release new versions of Rust libraries in general?

**Comment by johannesvollmer** [comment]
Furthermore I have done some more benchmarking. The results are a little bit worrying.

Compared to older releases, the last few optimizations (we profiled and changed code) produced a speedup of about 10% - 20%. However, just by upgrading to the newest version of half, we lose most of that performance again. I'll post the results of these benchmarks when I'm back at my computer again (not before Monday)

**Reply by johannesvollmer** [reply]
Here are the benchmark results:
```
2.0.0 read_f16_as_f16_uncompressed_1thread ... bench:  12,504,710 ns/iter (+/- 3,240,396)
1.7.0 read_f16_as_f16_uncompressed_1thread ... bench:  11,442,170 ns/iter (+/- 2,247,302)
1.6.3 read_f16_as_f16_uncompressed_1thread ... bench:  13,527,540 ns/iter (+/- 732,555)

2.0.0 read_f16_as_f16_zip_1thread          ... bench:  25,853,230 ns/iter (+/- 13,555,228)
1.7.0 read_f16_as_f16_zip_1thread          ... bench:  24,430,500 ns/iter (+/- 1,331,690)
1.6.3 read_f16_as_f16_zip_1thread          ... bench:  28,433,210 ns/iter (+/- 1,138,974)

2.0.0 read_f16_as_f16_zip_nthreads         ... bench:  14,045,660 ns/iter (+/- 2,492,090)
1.7.0 read_f16_as_f16_zip_nthreads         ... bench:  13,298,780 ns/iter (+/- 4,797,040)
1.6.3 read_f16_as_f16_zip_nthreads         ... bench:  14,484,800 ns/iter (+/- 4,977,872)

2.0.0 read_f16_as_f32_uncompressed_1thread ... bench:  12,216,300 ns/iter (+/- 6,689,122)
1.7.0 read_f16_as_f32_uncompressed_1thread ... bench:  11,440,550 ns/iter (+/- 655,714)
1.6.3 read_f16_as_f32_uncompressed_1thread ... bench:  13,572,920 ns/iter (+/- 901,939)

2.0.0 read_f16_as_f32_zip_1thread          ... bench:  25,749,090 ns/iter (+/- 1,331,989)
1.7.0 read_f16_as_f32_zip_1thread          ... bench:  24,614,090 ns/iter (+/- 2,303,655)
1.6.3 read_f16_as_f32_zip_1thread          ... bench:  28,501,120 ns/iter (+/- 3,182,333)

2.0.0 read_f16_as_f32_zip_nthreads         ... bench:  16,218,520 ns/iter (+/- 8,774,510)
1.7.0 read_f16_as_f32_zip_nthreads         ... bench:  13,185,070 ns/iter (+/- 1,219,058)
1.6.3 read_f16_as_f32_zip_nthreads         ... bench:  14,331,860 ns/iter (+/- 1,563,790)

2.0.0 read_f16_as_u32_uncompressed_1thread ... bench:  11,498,780 ns/iter (+/- 980,249)
1.7.0 read_f16_as_u32_uncompressed_1thread ... bench:  11,558,720 ns/iter (+/- 2,058,078)
1.6.3 read_f16_as_u32_uncompressed_1thread ... bench:  13,478,500 ns/iter (+/- 633,129)

2.0.0 read_f32_as_f16_uncompressed_1thread ... bench:  42,801,690 ns/iter (+/- 16,234,217)
1.7.0 read_f32_as_f16_uncompressed_1thread ... bench:  42,234,700 ns/iter (+/- 1,684,505)
1.6.3 read_f32_as_f16_uncompressed_1thread ... bench:  41,605,070 ns/iter (+/- 1,971,679)

2.0.0 read_f32_as_f16_zips_1thread         ... bench: 122,554,780 ns/iter (+/- 38,006,974)
1.7.0 read_f32_as_f16_zips_1thread         ... bench: 120,550,690 ns/iter (+/- 6,430,166)
1.6.3 read_f32_as_f16_zips_1thread         ... bench: 124,590,170 ns/iter (+/- 14,099,835)

2.0.0 read_f32_as_f16_zips_nthreads        ... bench:  60,497,930 ns/iter (+/- 29,453,111)
1.7.0 read_f32_as_f16_zips_nthreads        ... bench:  58,007,700 ns/iter (+/- 7,720,776)
1.6.3 read_f32_as_f16_zips_nthreads        ... bench:  58,505,830 ns/iter (+/- 15,191,091)

2.0.0 read_f32_as_f32_uncompressed_1thread ... bench:  43,541,660 ns/iter (+/- 18,622,460)
1.7.0 read_f32_as_f32_uncompressed_1thread ... bench:  42,621,910 ns/iter (+/- 3,484,493)
1.6.3 read_f32_as_f32_uncompressed_1thread ... bench:  41,796,660 ns/iter (+/- 1,725,040)

2.0.0 read_f32_as_f32_zips_1thread         ... bench: 125,346,950 ns/iter (+/- 41,740,642)
1.7.0 read_f32_as_f32_zips_1thread         ... bench: 121,884,960 ns/iter (+/- 9,169,102)
1.6.3 read_f32_as_f32_zips_1thread         ... bench: 129,178,550 ns/iter (+/- 13,316,515)

2.0.0 read_f32_as_f32_zips_nthreads        ... bench:  61,147,750 ns/iter (+/- 28,670,965)
1.7.0 read_f32_as_f32_zips_nthreads        ... bench:  58,508,860 ns/iter (+/- 20,310,249)
1.6.3 read_f32_as_f32_zips_nthreads        ... bench:  67,263,550 ns/iter (+/- 11,064,945)

2.0.0 read_f32_as_u32_uncompressed_1thread ... bench:  46,673,270 ns/iter (+/- 22,380,346)
1.7.0 read_f32_as_u32_uncompressed_1thread ... bench:  42,306,660 ns/iter (+/- 3,177,281)
1.6.3 read_f32_as_u32_uncompressed_1thread ... bench:  44,077,260 ns/iter (+/- 8,076,883)


2.0.0 write_nonparallel_zip1_to_buffered      ... bench: 570,570,670 ns/iter (+/- 127,699,252)
1.7.0 write_nonparallel_zip1_to_buffered      ... bench: 422,853,500 ns/iter (+/- 18,556,625)
1.6.3 write_nonparallel_zip1_to_buffered      ... bench: 494,333,310 ns/iter (+/- 22,001,115)

2.0.0 write_parallel_any_channels_to_buffered ... bench:  45,503,670 ns/iter (+/- 14,305,588)
1.7.0 write_parallel_any_channels_to_buffered ... bench:  41,496,310 ns/iter (+/- 4,816,822)
1.6.3 write_parallel_any_channels_to_buffered ... bench:  43,039,850 ns/iter (+/- 6,130,950)

2.0.0 write_parallel_zip16_to_buffered        ... bench: 134,109,520 ns/iter (+/- 18,715,275)
1.7.0 write_parallel_zip16_to_buffered        ... bench: 118,666,500 ns/iter (+/- 8,939,185)
1.6.3 write_parallel_zip16_to_buffered        ... bench: 139,027,520 ns/iter (+/- 12,985,769)

2.0.0 write_parallel_zip1_to_buffered         ... bench: 122,695,370 ns/iter (+/- 14,848,035)
1.7.0 write_parallel_zip1_to_buffered         ... bench: 102,092,290 ns/iter (+/- 9,285,430)
1.6.3 write_parallel_zip1_to_buffered         ... bench: 124,118,150 ns/iter (+/- 10,104,474)

2.0.0 write_uncompressed_to_buffered          ... bench:  28,015,560 ns/iter (+/- 7,693,241)
1.7.0 write_uncompressed_to_buffered          ... bench:  27,416,260 ns/iter (+/- 3,018,483)
1.6.3 write_uncompressed_to_buffered          ... bench:  28,745,010 ns/iter (+/- 3,703,883)
```

where `2.0.0` represents the same as `1.7.0`, but with `half` increased to version `2.3`

**Reply by Shnatsel** [reply]
Huh. `1.7.0` is `half 2.2` plus `use-intrinsics` feature, right?

**Reply by johannesvollmer** [reply]
No feature flags anywhere. So, no intrinsics except for `2.0.0`.

From `1.6.3` to `1.7.0`, there were various small exr optimizations, but nothing about half changed

**Reply by johannesvollmer** [reply]
It's really strange though, there must be another reason, because using the new `half` version degrades performance of benchmarks that don't even have any `f16` conversions in them! maybe code bloat led to cache misses in the code pages?

```
2.0.0 read_f32_as_u32_uncompressed_1thread ... bench:  46,673,270 ns/iter (+/- 22,380,346)
1.7.0 read_f32_as_u32_uncompressed_1thread ... bench:  42,306,660 ns/iter (+/- 3,177,281)
1.6.3 read_f32_as_u32_uncompressed_1thread ... bench:  44,077,260 ns/iter (+/- 8,076,883)
```

**Reply by Shnatsel** [reply]
Okay, first, the file used for `read_f32_as_f32` benchmarks must actually be an `f16` file, because on my machine it has the _exact same_ performance as `f16` conversions. It rises and falls exactly in tandem with the `f16` benchmarks, so it must be actually `f16`.

Second, I cannot reproduce the regression on my machine. Please make sure you're benchmarking with the same compiler version at all times. I tried `half` 2.2, 2.2 with `use-intrinsics` feature and 2.3 using `rustc 1.73.0-nightly (8ca44ef9c 2023-07-10)`. For me the results are exactly the same using 2.2 with `use-intrinsics` feature and 2.3, and both beat the basic 2.2.

**Reply by johannesvollmer** [reply]
Oh, this would explain it! I'll do it again and double check the files and compiler version, good idea :) how can I make sure the same compiler is used? isn't it enough to just do `rustup update` once?

**Comment by Shnatsel** [comment]
The crux of the issue is that `cargo` does not take MSRV into account, so it can upgrade to crates that no longer build on your current `rustc`. Making `cargo` do it is the only good solution. There is simply nothing the crate authors can do to make the problem go away.

There are options with various trade-offs, but all of them have issues:

1. Major semver bump
    -  Drawback: makes anything depending on 1.x and 2.x incompatible, requiring the entire ecosystem of dependent crates to upgrade in lock-step. A huge headache for everyone involved.
1. Minor version bump
    - Drawback: breaks builds on older Rust without a `Cargo.lock`, or when running `cargo update`, but at least a manual workaround is _possible._
1. Introduce an optional, default feature for using newer compiler features
    -  Drawback: same as (2). May not even be possible when transitive dependencies like `half` are involved.
1. Introduce an optional, non-default feature for using newer compiler features
    -  Drawback: Users on new compilers get sub-par performance without taking additional steps. May not even be possible when transitive dependencies like `half` are involved.
1. Wait until the required Rust version is reasonably widely used
    -  Drawback: improvements delayed by half a year or more for *everyone* just because of a minority on an older Rust version. Still doesn't make everyone happy.
1. Use build.rs to detect the Rust version and toggle features accordingly
    -  Drawback: users find this behavior surprising. Also, the presence of build.rs complicates integration into other build systems, required e.g. when calling Rust code from another language.

But without first-class MSRV support in Cargo this is always a losing battle. Even if you've opted for one of the solutions, some transitive dependency will issue a new release with a higher MSRV and screw up all your hard work, and then it turns out you've done all these trade-offs for nothing.

**Reply by johannesvollmer** [reply]
Thank you for this summary <3 I think I got the bigger picture now. Considering that the new release of `half` does not improve performance, but regress instead, I suggest the solution today is to wait (5). And possibly find out why performance regressed.

In general however, I feel like (2) would be the least exhausting and least error-prone way for library developers, even though it might cause users to investigate why their build failed after running `cargo update`.

**Reply by Shnatsel** [reply]
I also prefer (2)

**Comment by johannesvollmer** [comment]
The last plan was to release a new major version of `exr` (1.8.0, not 2.0.0) that allows the newest version of half. But the benchmarks show a regression. Maybe we should just leave it as-is?
